README.md
# catalauncher

This is a tool for helping you play [Cataclysm: Dark Days
Ahead](https://github.com/CleverRaven/Cataclysm-DDA/) (CDDA). CDDA is an open
source post-apocalyptic roguelike. It's a lot of fun, and this tool will make
it easier to play on Linux (and possible macOS and other Unix systems).

Please note that if you're on Windows there is a much more polished GUI
launcher, the [CDDA Game
Launcher](https://github.com/remyroy/CDDA-Game-Launcher).

## How to Use It

You can download a binary version of the latest build on this project's
[GitHub releases
page](https://github.com/houseabsolute/catalauncher/releases). Put this in a
directory in your path. Since this is released with
[goreleaser](https://goreleaser.com/) you can also use
[godownloader](https://github.com/goreleaser/godownloader) to install it if
you prefer.

You'll need [Docker CE installed](https://docs.docker.com/install/) in order
to play the game since it's launched inside a Docker container. See the linked
docs for information on installing Docker CE.

## Setup

Before launching the game you need to set the launcher up. This is mostly to
tell it where to store files:

```
$> catalauncher setup
```

This will ask where you want to store game files (and in the future may ask
more questions). By default files are stored under
`$HOME/.catalauncher`. Accepting this default will make your life a little
simpler. Otherwise you'll need to tell it where your config file lives every
time you run it.

## Launching

To launch the game simply run the `launch` subcommand:

```
$> catalauncher launch
```

## Options

* `--config` - The location of your config file. This is accepted by all
  subcommands. Note that if your config file is not in the default location,
  `$HOME/.catalauncher`, then you'll need to pass this every time you run the
  `launch `subcommand.
* `--build` - This is an option for the launch subcommand. Pass this to
  specify which build you'd like to launch. By default you always get the most
  recent build.

## How It Works and What It Does

When you run `launch` there a number of things that happen.

### Fetching New Builds

First, the launcher checks for a new binary build in
http://dev.narc.ro/cataclysm/jenkins-latest/Linux_x64/Tiles/. These builds are
created via Jenkins.

If there is a new build it will be downloaded and untarred (unless you asked
for an older build with the `--build` flag).

If the launcher is fetching a new build it will open the [Jenkins CDDA changes
list](http://gorgon.narc.ro:8080/job/Cataclysm-Matrix/changes) in your browser
so you can see what's new.

### Character Creation Templates

If your most recent local build has any character creation templates, those
are copied into the new build automatically (unfortunately these must live
under the game's directory, rather than in a shared location).

### Extras (Mods & Soundpacks)

The launcher also provides some mods and soundpacks. These are maintained in
the [houseabsolute/cataclysm-extras-collection
repo](https://github.com/houseabsolute/cataclysm-extras-collection). If you
want to add a mod, soundpack, or tileset, please [file a PR or issue
there](https://github.com/houseabsolute/cataclysm-extras-collection/issues). If
a mod is causing the game to crash, please [file a PR or issue to remove
it](https://github.com/houseabsolute/cataclysm-extras-collection/issues).

Whenever you run the launcher a local copy of that git repo is
pulled/updated. The contents are then copied into the per-build game directory
(unfortunately CDDA does not work when these directories are symlinked).

### Docker

The game itself is run in a Docker container using my
[houseabsolute/catalauncher-player](https://cloud.docker.com/u/houseabsolute/repository/docker/houseabsolute/catalauncher-player)
image. This image is built using the
[Dockerfile](https://github.com/houseabsolute/catalauncher/blob/master/docker/Dockerfile)
in this repo. This avoids the need to install any libraries on the host
system.

This container is run with quite a bit of access to the host system in order
to make video and sound work. I'm using Docker primarily for convenience
rather than isolation. Notably, the container is run with access to the
following files/directories/env vars on the host system:

* `/etc/machine-id`
* `/run/user/$USER_ID/pulse`
* `$HOME/.pulse`
* `/dev/dri`
* `/tmp/.X11-unix`
* `/var/lib/dbus`
* The `$DISPLAY` env var

However, The game will be executed using your user and group ids, not `root`.

Given this it's not clear to me whether this will work with Docker on macOS or
Windows (or even a Linux system that is very different from my own desktop
running Ubuntu 18.04). Patches to handle a greater variety of host systems are
welcome!

### Game Files

The launcher stores your game config, graveyard, and save files outside of the
per-build game directories, so they will persist as you run new builds.

## Todo and Known Issues

When your character dies (which will happen a lot) you'll see an error about
renaming some memorial files. I'm not sure why these are showing up. The
graveyard _does_ contain records of character deaths so it doesn't seem to be
a major error.

As mentioned above, the way Docker is run in order to provide it access to
video and sound on the host system greatly reduces container isolation and may
be very specific to my desktop.


Changes.md
## 0.0.6  2020-06-05

* Fix handling of extracted tarball so we find the game dir when the release
  letter changes.


## 0.0.5  2019-06-15

* Some config files still end up under the game's directory tree even when
  --configdir is passed. These are now copied to the new build when you
  download one.


## 0.0.4  2019-05-29

* Fixed parsing of build filenames.


## 0.0.3  2019-03-09

* Fixed the code for the change in Cataclysm version (now 0.D).


## 0.0.2  2019-01-26

* Added a clean subcommand to clean out old builds.


## 0.0.1  2019-01-20

* First release upon an unsuspecting world.


launcher/launcher.go
package launcher

import (
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"os/exec"
	"os/user"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"code.gitea.io/git"
	"github.com/PuerkitoBio/goquery"
	"github.com/houseabsolute/catalauncher/config"
	"github.com/houseabsolute/catalauncher/curuser"
	"github.com/houseabsolute/catalauncher/localbuilds"
	"github.com/houseabsolute/catalauncher/util"
	"github.com/otiai10/copy"
	"github.com/skratchdot/open-golang/open"
	pb "gopkg.in/cheggaaa/pb.v2"
)

type build struct {
	uri         string
	filename    string
	version     string
	buildNumber uint
	date        time.Time
}

type Launcher struct {
	config      *config.Config
	local       *localbuilds.LocalBuilds
	build       uint
	user        *curuser.User
	stdout      io.Writer
	stderr      io.Writer
	buildsURI   string
	currentUser *user.User
}

const defaultBuildsURI = "http://dev.narc.ro/cataclysm/jenkins-latest/Linux_x64/Tiles/"

func New(rootDir string, build uint) (*Launcher, error) {
	c, err := config.New(rootDir)
	if err != nil {
		return nil, err
	}

	user, err := curuser.New()
	if err != nil {
		return nil, err
	}

	return &Launcher{
		config:    c,
		local:     localbuilds.New(c),
		build:     build,
		user:      user,
		stdout:    os.Stdout,
		stderr:    os.Stderr,
		buildsURI: defaultBuildsURI,
	}, nil
}

func (l *Launcher) Launch() error {
	wanted, err := l.determineWantedBuild()
	if err != nil {
		return err
	}

	localLatest, err := l.local.Latest()
	if err != nil {
		return err
	}

	exists, err := l.local.HasBuild(wanted.buildNumber)
	if err != nil {
		return err
	}

	if !exists {
		err := l.downloadBuild(wanted)
		if err != nil {
			return err
		}
		if localLatest != 0 {
			err = l.copyTemplates(localLatest, wanted.buildNumber)
			if err != nil {
				return err
			}

			err = l.copyGameConfig(localLatest, wanted.buildNumber)
			if err != nil {
				return err
			}
		}
	}

	err = l.makeGameConfigDir(wanted)
	if err != nil {
		return err
	}

	err = l.updateExtras(wanted)
	if err != nil {
		return err
	}

	err = l.pullDockerImage()
	if err != nil {
		return err
	}

	return l.launchGame(wanted)
}

func (l *Launcher) determineWantedBuild() (build, error) {
	if l.build == 0 {
		return l.latestBuild()
	}

	builds, err := l.parseBuilds()
	if err != nil {
		return build{}, err
	}

	var wanted build
	for _, b := range builds {
		if b.buildNumber == l.build {
			wanted = b
			break
		}
	}
	if wanted.uri == "" {
		return build{}, fmt.Errorf(
			"Could not find the build you requested, #%d, in the list of available builds", l.build)
	}

	return wanted, nil
}

func (l *Launcher) latestBuild() (build, error) {
	builds, err := l.parseBuilds()
	if err != nil {
		return build{}, err
	}

	if len(builds) == 0 {
		return build{}, errors.New("Could not find any builds!")
	}

	util.Say(l.stdout, "Found %d builds", len(builds))

	localLatest, err := l.local.Latest()
	if err != nil {
		return build{}, err
	}

	if localLatest == 0 {
		util.Say(l.stdout, "No builds have been downloaded yet")
	} else if localLatest != builds[0].buildNumber {
		util.Say(l.stdout, "Latest local build is #%d", localLatest)
		util.Say(
			l.stdout,
			"The latest build is build #%d, released %s",
			builds[0].buildNumber, builds[0].date.Format("2006-01-02 15:04"),
		)
	} else {
		util.Say(
			l.stdout,
			"You have the latest build, #%d, released %s",
			builds[0].buildNumber, builds[0].date.Format("2006-01-02 15:04"),
		)
	}

	return builds[0], nil
}

var fileRE = regexp.MustCompile(`^cataclysmdda-([0-9].[A-Z]-Linux_x64-Tiles-(\d+))\.tar\.gz$`)

func (l *Launcher) parseBuilds() ([]build, error) {
	util.Say(l.stdout, "Getting list of builds from %s", l.buildsURI)
	res, err := http.Get(l.buildsURI)
	if err != nil {
		return []build{}, fmt.Errorf("Could not fetch build list from %s: %s", l.buildsURI, err)
	}
	defer res.Body.Close()
	if res.StatusCode != 200 {
		return []build{}, fmt.Errorf(
			"Did not get a 200 status when fetching %s, got a %d (%s) instead",
			l.buildsURI, res.StatusCode, res.Status,
		)
	}

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return []build{}, fmt.Errorf("Error parsing HTML from %s: %s", l.buildsURI, err)
	}

	buildDates, err := l.parseBuildDates(doc)
	if err != nil {
		return []build{}, err
	}

	builds := []build{}
	var eachErr error
	doc.Find("a").Each(func(_ int, sel *goquery.Selection) {
		if eachErr != nil {
			return
		}

		href, _ := sel.Attr("href")
		m := fileRE.FindStringSubmatch(href)
		if len(m) < 2 {
			return
		}

		num, err := strconv.Atoi(m[2])
		if err != nil {
			eachErr = fmt.Errorf("Could not convert %s to an integer: %s", m[2], err)
		}

		builds = append(
			builds,
			build{
				uri:         l.buildsURI + href,
				filename:    href,
				version:     m[1],
				buildNumber: uint(num),
				date:        buildDates[href],
			},
		)
	})
	if eachErr != nil {
		return []build{}, eachErr
	}

	// Using After gives us a reverse sorting from most to least recent.
	sort.SliceStable(builds, func(i, j int) bool { return builds[i].date.After(builds[j].date) })
	return builds, nil
}

var buildDatesRE = regexp.MustCompile(`(cataclysmdda-\S+\.tar\.gz)\s+(2\d\d\d-\d\d-\d\d \d\d:\d\d)`)

func (l *Launcher) parseBuildDates(doc *goquery.Document) (map[string]time.Time, error) {
	dates := map[string]time.Time{}
	m := buildDatesRE.FindAllStringSubmatch(doc.Find("body").First().Text(), -1)
	for _, pair := range m {
		d, err := time.Parse("2006-01-02 15:04", pair[2])
		if err != nil {
			return dates, fmt.Errorf("Could not parse date for the file %s from text (%s)", pair[1], pair[2])
		}
		dates[pair[1]] = d
	}
	return dates, nil
}

const changesURI = "http://gorgon.narc.ro:8080/job/Cataclysm-Matrix/changes"

func (l *Launcher) downloadBuild(b build) error {
	util.Say(l.stdout, "Downloading build #%d from %s", b.buildNumber, b.uri)
	util.Say(l.stdout, "Opening the changes listing in your browser")
	open.Start(changesURI)

	dir, err := ioutil.TempDir("", "catalauncher-")
	if err != nil {
		return fmt.Errorf("Could not create a temporary directory: %s", err)
	}

	file := filepath.Join(dir, b.filename)
	out, err := os.Create(file)
	if err != nil {
		return fmt.Errorf("Could not create file at %s: %s", file, err)
	}
	defer out.Close()

	resp, err := http.Get(b.uri)
	if err != nil {
		return fmt.Errorf("Could not get %s: %s", b.uri, err)
	}
	defer resp.Body.Close()

	cl := resp.Header.Get("Content-Length")
	len := 0
	if cl != "" {
		len, err = strconv.Atoi(cl)
		if err != nil {
			return fmt.Errorf("Could not convert %s to an integer: %s", cl, err)
		}
	}

	bar := pb.New(len)
	bar.Start()
	rd := bar.NewProxyReader(resp.Body)

	_, err = io.Copy(out, rd)
	if err != nil {
		return fmt.Errorf("Could not save %s to %s: %s", b.uri, file, err)
	}

	return l.untarBuild(file, b)
}

func (l *Launcher) untarBuild(file string, b build) error {
	target := filepath.Join(l.config.BuildsDir(), fmt.Sprintf("%d", b.buildNumber))
	err := l.mkdir(target)
	if err != nil {
		return err
	}

	cmd := exec.Command("tar", "xzf", file, "-C", target)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf(`Could not run "tar xvzf %s -C %s": %s\n%s`, file, target, err, out)
	}

	return nil
}

func (l *Launcher) copyTemplates(from, to uint) error {
	return l.rcopy(
		filepath.Join(l.config.GameDir(from), "templates"),
		filepath.Join(l.config.GameDir(to), "templates"),
		"template",
	)
}

// Some config files end up in the game dir even when you pass
// --configdir. Why?
func (l *Launcher) copyGameConfig(from, to uint) error {
	fromDir := filepath.Join(l.config.GameDir(from), "config")
	_, err := os.Stat(fromDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}

	return l.rcopy(
		fromDir,
		filepath.Join(l.config.GameDir(to), "config"),
		"config",
	)
}

func (l *Launcher) makeGameConfigDir(b build) error {
	return os.MkdirAll(filepath.Join(l.config.GameDir(b.buildNumber), "config"), 0755)
}

const extrasGitRepo = "https://github.com/houseabsolute/cataclysm-extras-collection.git"

func (l *Launcher) updateExtras(b build) error {
	err := l.mkdir(l.config.ExtrasDir())
	if err != nil {
		return err
	}

	_, err = os.Stat(filepath.Join(l.config.ExtrasDir(), ".git"))
	if err != nil {
		if !os.IsNotExist(err) {
			return err
		}
		util.Say(l.stdout, "Cloning extras from %s", extrasGitRepo)
		err = git.Clone(extrasGitRepo, l.config.ExtrasDir(), git.CloneRepoOptions{})
		if err != nil {
			return err
		}
	} else {
		util.Say(l.stdout, "Updating extras git repo")
		err = git.Pull(l.config.ExtrasDir(), git.PullRemoteOptions{Remote: "origin"})
		if err != nil {
			return err
		}
	}

	things := []struct {
		from      string
		to        string
		what      string
		underData bool
	}{
		{"gfx", "gfx", "tileset", false},
		{"mods", "mods", "mod", true},
		{"soundpacks", "sound", "soundpack", true},
	}
	for _, t := range things {
		toElt := []string{l.config.GameDir(b.buildNumber)}
		if t.underData {
			toElt = append(toElt, "data")
		}
		toElt = append(toElt, t.to)
		err = l.rcopy(
			filepath.Join(l.config.ExtrasDir(), t.from),
			filepath.Join(toElt...),
			t.what,
		)
		if err != nil {
			return err
		}
	}

	return nil
}

func (l *Launcher) rcopy(from, to, what string) error {
	dir, err := os.Open(from)
	if err != nil {
		return err
	}

	entries, err := dir.Readdir(-1)
	if err != nil {
		return err
	}

	for _, e := range entries {
		if e.Name() == ".git" {
			continue
		}
		util.Say(l.stdout, "Copying %s %s to game dir", e.Name(), what)
		err := copy.Copy(filepath.Join(from, e.Name()), filepath.Join(to, e.Name()))
		if err != nil {
			return err
		}
	}

	return nil
}

func (l *Launcher) pullDockerImage() error {
	util.Say(l.stdout, "Pulling the latest houseabsolute/cataplayer-launcher image")
	return l.runCommand("docker", []string{"pull", "houseabsolute/catalauncher-player"})
}

func (l *Launcher) launchGame(b build) error {
	dataDir := l.config.GameDataDir()
	err := l.mkdir(dataDir)
	if err != nil {
		return err
	}

	runPulse := fmt.Sprintf("/run/user/%s/pulse", l.user.Uid)

	args := []string{
		"run",
		// We don't want the container sticking around once the game exits.
		"--rm",
		// We want to make sure save files and such are owned by the current
		// user, not root.
		"--user", fmt.Sprintf("%s:%s", l.user.Uid, l.user.Gid),
		// Needed for sound w/ Pulseaudio
		"-v", "/etc/machine-id:/etc/machine-id",
		"-v", runPulse + ":" + runPulse,
		"-v", "/var/lib/dbus:/var/lib/dbus",
		"-v", fmt.Sprintf("%s/.pulse:/.pulse", l.user.HomeDir),
		// Needed for graphics
		"-e", "DISPLAY",
		"--device", "/dev/dri",
		"-v", "/tmp/.X11-unix:/tmp/.X11-unix",
		//
		"-v", dataDir + ":/data",
		"-v", l.config.GameDir(b.buildNumber) + ":/game",
		// CDDA seems to expect PWD to be the game root dir.
		"-w", "/game",
		"houseabsolute/catalauncher-player:latest",
		"./cataclysm-tiles",
		"--savedir", "/data/save/",
		"--configdir", "/data/config/",
		"--memorialdir", "/data/graveyard/",
	}

	err = l.runCommand("docker", args)
	if err != nil {
		return err
	}
	os.Exit(0)

	// We should never get here for obvious reasons
	return nil
}

func (l *Launcher) runCommand(exe string, args []string) error {
	cmd := exec.Command(exe, args...)
	out, err := cmd.CombinedOutput()
	if err != nil {
		msg := fmt.Sprintf("Could not run \"%s %s\": %s\n", exe, strings.Join(args, " "), err)
		if len(out) > 0 {
			msg += "\n" + string(out)
		}
		return errors.New(msg)
	}

	return nil
}

func (l *Launcher) mkdir(dir string) error {
	err := os.MkdirAll(dir, 0755)
	if err != nil {
		return fmt.Errorf("Could not make directory %s: %s", dir, err)
	}
	return nil
}
